Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,312 +41,588 @@ public sealed class RCDSystem : EntitySystem
+                    _adminLogger.Add(LogType.RCD, LogImpact.High, $"{ToPrettyString(user):user} used RCD to set grid: {mapGridData.GridUid} tile: {mapGridData.Position} open to space");
                 }
-                if (IsTileBlocked(tile))
+                else
                 {
-                    _popup.PopupClient(Loc.GetString("rcd-component-tile-obstructed-message"), uid, user);
-                    return false;
+                    // Deconstruct object
+                    _adminLogger.Add(LogType.RCD, LogImpact.High, $"{ToPrettyString(user):user} used RCD to delete {ToPrettyString(target):target}");
+                    QueueDel(target);
                 }
-                return true;
-            default:
-                return false; //I don't know why this would happen, but sure I guess. Get out of here invalid state!
+
+                break;
         }
     }
 
-    private void NextMode(EntityUid uid, RCDComponent comp, EntityUid user)
+    #endregion
+
+    #region Utility functions
+
+    public bool TryGetMapGridData(EntityCoordinates location, [NotNullWhen(true)] out MapGridData? mapGridData)
     {
-        _audio.PlayPredicted(comp.SwapModeSound, uid, user);
+        mapGridData = null;
+        var gridUid = location.GetGridUid(EntityManager);
+
+        if (!TryComp<MapGridComponent>(gridUid, out var mapGrid))
+        {
+            location = location.AlignWithClosestGridTile(1.75f, EntityManager);
+            gridUid = location.GetGridUid(EntityManager);
+
+            // Check if we got a grid ID the second time round
+            if (!TryComp(gridUid, out mapGrid))
+                return false;
+        }
+
+        gridUid = mapGrid.Owner;
 
-        var mode = (int) comp.Mode;
-        mode = ++mode % RcdModeCount;
-        comp.Mode = (RcdMode) mode;
-        Dirty(uid, comp);
+        var tile = _mapSystem.GetTileRef(gridUid.Value, mapGrid, location);
+        var position = _mapSystem.TileIndicesFor(gridUid.Value, mapGrid, location);
+        mapGridData = new MapGridData(gridUid.Value, mapGrid, location, tile, position);
 
-        var msg = Loc.GetString("rcd-component-change-mode", ("mode", comp.Mode.ToString()));
-        _popup.PopupClient(msg, uid, user);
+        return true;
     }
 
-    private bool IsTileBlocked(TileRef tile)
+    public bool DoesCustomBoundsIntersectWithFixture(Box2 boundingBox, Angle boundingBoxAngle, EntityUid fixtureOwner, Fixture fixture)
     {
-        return _turf.IsTileBlocked(tile, CollisionGroup.MobMask);
+        var entXformComp = Transform(fixtureOwner);
+        var entXform = new Transform(entXformComp.LocalPosition, entXformComp.LocalRotation);
+
+        var verts = new ValueList<Vector2>()
+        {
+            boundingBox.BottomLeft,
+            boundingBox.BottomRight,
+            boundingBox.TopRight,
+            boundingBox.TopLeft
+        };
+
+        var poly = new PolygonShape();
+        poly.Set(verts.Span, 4);
+
```

Comment:
PolygonShape has box setters + can just store this on the system as a field to avoid re-allocating it too.